pkg/tcpip/transport/tcp/connect.go

```diff
@@ -1088,6 +1099,10 @@ func (e *endpoint) protocolMainLoop(handshake bool) *tcpip.Error {
 	// Mark endpoint as closed.
 	e.mu.Lock()
 	if e.state != StateError {
+		if e.state == StateEstablished || e.state == StateCloseWait {
+			e.stack.Stats().TCP.EstablishedResets.Increment()
+			e.stack.Stats().TCP.CurrentEstablished.Decrement()
+		}
 		e.state = StateClose
```

Comment:
The connection cannot be in Established state if it reaches here. Same for StateCloseWait. I believe the 2 states that are possible at this point are 

a) StateTimeWait 
b) StateClose

The code here transitions to StateClose if it receives the last ACK properly https://github.com/google/gvisor/blob/816a9211e96b3eabe7bead7fa625831fe8d2fbf8/pkg/tcpip/transport/tcp/rcv.go#L215

Reply:
Make sense. I did not and do not understand the state machine in netstack, which introduces some unnecessary changes. Thank you for the review and suggestions!

Reply:
Not done yet?